nineml/abstraction/expressions/base.py

```diff
@@ -180,43 +181,7 @@ def rhs_atoms(self):
     def rhs_as_python_func(self):
         """ Returns a python callable which evaluates the expression in
         namespace and returns the result """
-        def nineml_expression(**kwargs):
-            if isinstance(self.rhs, (bool, int, float, BooleanTrue,
-                                     BooleanFalse)):
-                val = self.rhs
-            else:
-                if self.rhs.is_Boolean:
-                    try:
-                        val = self.rhs.subs(kwargs)
-                    except Exception:
-                        raise NineMLUsageError(
-                            "Incorrect arguments provided to expression ('{}')"
-                            ": '{}'\n".format(
-                                "', '".join(self.rhs_symbol_names),
-                                "', '".join(list(kwargs.keys()))))
-                else:
-                    try:
-                        val = self.rhs.evalf(subs=kwargs)
-                    except Exception:
-                        raise NineMLUsageError(
-                            "Incorrect arguments provided to expression '{}'"
-                            ": '{}' (expected '{}')\n".format(
-                                self.rhs,
-                                "', '".join(list(kwargs.keys())),
-                                "', '".join(self.rhs_symbol_names)))
-                    try:
-                        val = float(val)
-                    except TypeError:
-                        try:
-                            locals_dict = deepcopy(kwargs)
-                            locals_dict.update(str_to_npfunc_map)
-                            val = eval(str(val), {}, locals_dict)
-                        except Exception:
-                            raise NineMLUsageError(
-                                "Could not evaluate expression: {}"
-                                .format(self.rhs_str))
-            return val
-        return nineml_expression
+        return lambdify(list(self.rhs_symbol_names), self.rhs, 'numpy')
```

Comment:
does this single line really achieve the same as the deleted code?

Reply:
Yes, and runs much faster for complex expressions as it uses Numpy under the hood :).

I think I must have missed `lambdify` when I was converting the library to Sympy, either that or it was added later. I have been using it in my `ref_implementation` branch so I am fairly familiar with how it works now. Also, if you have a look at that code it mostly just uses the `evalf` method but then has to handle the special cases when the expression is a relational (i.e. in Triggers) or has a built-in function in it (the `eval` at the bottom). Using Numpy avoids having to do all these things.

I had also been avoiding Numpy as a dependency as I liked the idea that the library was pure Python and then could be run in other languages (e.g. Java, Iron Python). But that is probably a very marginal gain and not worth the ease of handling array data with Numpy.